sycl/doc/design/DeviceIf.md

```diff
@@ -0,0 +1,277 @@
+# Implementation design for "device\_if"
+
+This document describes the design for the DPC++ implementation of the
+[sycl\_ext\_oneapi\_device\_if][1] extension, which allows device code to
+conditionally execute code based on the current device.
+
+[1]: <../extensions/proposed/sycl_ext_oneapi_device_if.asciidoc>
+
+
+## Phased implementation
+
+The implementation is divided into two phases.  In the first phase, we support
+only AOT mode.  The second phase adds support also for JIT mode.
+
+
```

Comment:
Can we add a section where all targets are summarized for simpler reference, and all other section would refer to it?
```suggestion
## Target device identification.

A number of sections in this document refer to device targets in various
contexts. This section provides a table with complete list of the targets
supported by this design, where a row corresponds to a specific target, and
columns define:

Description
  Description of the target.
AOT compiler ID
  How the target is specified in the AOT compiler invocation in the
  -fsycl-targets= option.
AOT compiler ID alias
  Alternative AOT compiler ID (alias) for some of the targets.
Aspect
  Specific enum id of sycl::aspect corresponding to the target.
Aspect alias
  Aleternative aspec enum id (with the same value as main enum id)
Preprocessor Macro
  Preprocessor macro (device compiler only) predefined when AOT-compiling
  for the target.


|AOT compiler ID  |AOT compiler ID alias |Aspect               | Aspect  alias       |Preprocessor Macro                    |Description                                     |
|-----------------|----------------------|---------------------|---------------------|--------------------------------------|------------------------------------------------|
|ptx64            |                      |                     |                     |                                      |Generic 64-bit PTX target architecture          |
|spir64           |                      |                     |                     |                                      |Generic 64-bit SPIR-V target                    |
|x86_64           |                      |ext_intel_x86_64     |                     |                                      |Any CPU device with the x86_64 instruction set. |
|intel_gpu_pvc    |                      |ext_intel_gpu_pvc    |                     |`__SYCL_TARGET_INTEL_GPU_PVC__`    |Ponte Vecchio Intel graphics architecture       |
|intel_gpu_acm_g12|                      |ext_intel_gpu_acm_g12|                     |`__SYCL_TARGET_INTEL_GPU_ACM_G12__`|Alchemist G12 Intel graphics architecture       |
|intel_gpu_acm_g11|                      |ext_intel_gpu_acm_g11|                     |`__SYCL_TARGET_INTEL_GPU_ACM_G11__`|Alchemist G11 Intel graphics architecture       |
|intel_gpu_acm_g10|                      |ext_intel_gpu_acm_g10|                     |`__SYCL_TARGET_INTEL_GPU_ACM_G10__`|Alchemist G10 Intel graphics architecture       |
|intel_gpu_dg1    |intel_gpu_12_10_0     |ext_intel_gpu_dg1    |ext_intel_gpu_12_10_0|`__SYCL_TARGET_INTEL_GPU_DG1__`    |DG1           Intel graphics architecture       |
|intel_gpu_adl_n  |                      |ext_intel_gpu_adl_n  |                     |`__SYCL_TARGET_INTEL_GPU_ADL_N__`  |Alder Lake N  Intel graphics architecture       |
|intel_gpu_adl_p  |                      |ext_intel_gpu_adl_p  |                     |`__SYCL_TARGET_INTEL_GPU_ADL_P__`  |Alder Lake P  Intel graphics architecture       |
|intel_gpu_rpl_s  |                      |ext_intel_gpu_rpl_s  |                     |`__SYCL_TARGET_INTEL_GPU_RPL_S__`  |Raptor Lake   Intel graphics architecture       |
|intel_gpu_adl_s  |                      |ext_intel_gpu_adl_s  |                     |`__SYCL_TARGET_INTEL_GPU_ADL_S__`  |Alder Lake S  Intel graphics architecture       |
|intel_gpu_rkl    |                      |ext_intel_gpu_rkl    |                     |`__SYCL_TARGET_INTEL_GPU_RKL__`    |Rocket Lake   Intel graphics architecture       |
|intel_gpu_tgllp  | intel_gpu_12_0_0     |ext_intel_gpu_tgllp  |ext_intel_gpu_12_0_0 |`__SYCL_TARGET_INTEL_GPU_TGLLP__`  |Tiger Lake    Intel graphics architecture       |
|intel_gpu_ehl    | intel_gpu_11_2_0     |ext_intel_gpu_ehl    |ext_intel_gpu_11_2_0 |`__SYCL_TARGET_INTEL_GPU_EHL__`    |Elkhart Lake  Intel graphics architecture       |
|intel_gpu_icllp  | intel_gpu_11_0_0     |ext_intel_gpu_icllp  |ext_intel_gpu_11_0_0 |`__SYCL_TARGET_INTEL_GPU_ICLLP__`  |Ice Lake      Intel graphics architecture       |
|intel_gpu_cml    |  intel_gpu_9_7_0     |ext_intel_gpu_cml    | ext_intel_gpu_9_7_0 |`__SYCL_TARGET_INTEL_GPU_CML__`    |Comet Lake    Intel graphics architecture       |
|intel_gpu_aml    |  intel_gpu_9_6_0     |ext_intel_gpu_aml    | ext_intel_gpu_9_6_0 |`__SYCL_TARGET_INTEL_GPU_AML__`    |Amber Lake    Intel graphics architecture       |
|intel_gpu_whl    |  intel_gpu_9_5_0     |ext_intel_gpu_whl    | ext_intel_gpu_9_5_0 |`__SYCL_TARGET_INTEL_GPU_WHL__`    |Whiskey Lake  Intel graphics architecture       |
|intel_gpu_glk    |  intel_gpu_9_4_0     |ext_intel_gpu_glk    | ext_intel_gpu_9_4_0 |`__SYCL_TARGET_INTEL_GPU_GLK__`    |Gemini Lake   Intel graphics architecture       |
|intel_gpu_apl    |  intel_gpu_9_3_0     |ext_intel_gpu_apl    | ext_intel_gpu_9_3_0 |`__SYCL_TARGET_INTEL_GPU_APL__`    |Apollo Lake   Intel graphics architecture       |
|intel_gpu_cfl    |  intel_gpu_9_2_9     |ext_intel_gpu_cfl    | ext_intel_gpu_9_2_9 |`__SYCL_TARGET_INTEL_GPU_CFL__`    |Coffee Lake   Intel graphics architecture       |
|intel_gpu_kbl    |  intel_gpu_9_1_9     |ext_intel_gpu_kbl    | ext_intel_gpu_9_1_9 |`__SYCL_TARGET_INTEL_GPU_KBL__`    |Kaby Lake     Intel graphics architecture       |
|intel_gpu_skl    |  intel_gpu_9_0_9     |ext_intel_gpu_skl    | ext_intel_gpu_9_0_9 |`__SYCL_TARGET_INTEL_GPU_SKL__`    |Skylake       Intel graphics architecture       |
|intel_gpu_bdw    |  intel_gpu_8_0_0     |ext_intel_gpu_bdw    | ext_intel_gpu_8_0_0 |`__SYCL_TARGET_INTEL_GPU_BDW__`    |Broadwell     Intel graphics architecture       |
```

Reply:
I'm not ignoring this, but I wanted to resolve @Pennycook's issue about the aspect vs. architecture enums first.

Note, however, that there isn't any duplication in the lists currently in this document, so adding a new section like this doesn't reduce any verbosity.

Reply:
Yes, there is no duplication, but my point is that e.g. to match e.g. `AOT compiler ID` to the corresponding `Preprocessor macro` one needs to scroll back and forth.
I'm not insisting, this is just how I would prefer this.

Reply:
I have a weak preference not to combine these into a single table.  The names for the `-fsycl-targets` option are relevant to both phase 1 and 2.  However, I think the macros like `__SYCL_TARGET_INTEL_GPU_BDW__` will be specific to phase 2.  Separating them will make it easier to update this document once we implement phase 2.